crates/starknet-devnet-core/src/state/mod.rs

```diff
@@ -55,72 +61,146 @@ pub trait CustomState {
 }
 
 #[derive(Default, Clone)]
-/// Utility structure that makes it easier to calculate state diff later on
+/// Utility structure that makes it easier to calculate state diff later on. Classes are first
+/// inserted into the staging area (pending state), later to be committed (latest state).
 pub struct CommittedClassStorage {
     staging: HashMap<ClassHash, ContractClass>,
-    committed: HashMap<ClassHash, ContractClass>,
+    committed: HashMap<ClassHash, (ContractClass, u64)>,
+    /// Remembers all classes committed at a block
+    block_number_to_classes: HashMap<u64, Vec<ClassHash>>,
 }
 
 impl CommittedClassStorage {
+    /// Insert a new class into the staging area. Once it can be committed, call `commit`.
     pub fn insert(&mut self, class_hash: ClassHash, contract_class: ContractClass) {
         self.staging.insert(class_hash, contract_class);
     }
 
-    pub fn commit(&mut self) -> HashMap<ClassHash, ContractClass> {
-        let diff = self.staging.clone();
-        self.committed.extend(self.staging.drain());
-        diff
+    /// Commits all of the staged classes and returns them, together with their hashes.
+    pub fn commit(&mut self, block_number: u64) -> HashMap<ClassHash, ContractClass> {
+        let mut freshly_committed = HashMap::new();
+
+        let hashes_at_this_block = self.block_number_to_classes.entry(block_number).or_default();
+        for (class_hash, class) in &self.staging {
+            freshly_committed.insert(*class_hash, class.clone());
+            self.committed.insert(*class_hash, (class.clone(), block_number));
+
+            hashes_at_this_block.push(*class_hash);
+        }
+
+        self.empty_staging();
+        freshly_committed
+    }
+
+    /// Returns sierra for cairo1; returns the only artifact for cairo0.
+    pub fn get_class(
+        &self,
+        class_hash: &ClassHash,
+        block_id: &BlockTagOrNumber,
```

Comment:
here is block_id but type is BlockTagOrNumber

Reply:
well it's still block id; I don't plan on calling it block_tag_or_number

Reply:
well in the whole project when we have variables called `block_id` we refer to BlockId enum, when we use `block_tag` we refer to `BlockTag`, when we use `block_number` we refer to `u64` and when we use 
 `block_hash` we refer to BlockHash
 
 so my expectation here is to have `block_tag_number`, or something like this to be consistent with the whole project

Reply:
but I don't wanna escalate Parkinson's law here since there are more important stuff to do, feel free to resolve it if you want

Reply:
Addressed in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/499/commits/7b61c28c85491cc5607b69340f2378de2ef5d53c